civihr_employee_portal/src/View/ViewsResultDiff.php

```diff
@@ -0,0 +1,284 @@
+   *
+   * @var array
+   */
+  public $oldData = [];
+
+  /**
+   * the difference between $newData and $oldData once comparison is made
+   *
+   * @var array
+   */
+  public $changes = [];
+
+  /**
+   * ViewsResultDiff constructor
+   * initialises viewSettings property
+   *
+   * @param array $viewsSettings
+   *     to initialize viewSettings property
+   */
+  function __construct($viewsSettings = []) {
+    // Normalising view settings structure
+    foreach ($viewsSettings as $index=>&$value ) {
+      if (isset($value['view'])) {
+        $value['args'] = !empty($value['args']) ? $value['args'] : [];
+        $value['displayId'] = !empty($value['displayId']) ? $value['displayId'] : 'default';
+      } else if (isset($value['group'])) {
+        foreach ($value['group'] as &$viewSetting) {
+          $viewSetting['args'] = !empty($viewSetting['args']) ? $viewSetting['args'] : [];
+          $viewSetting['displayId'] = !empty($viewSetting['displayId']) ? $viewSetting['displayId'] : 'default';
+        }
+      }
+    }
+    $this->viewsSettings = $viewsSettings;
+  }
+
+  /**
+   * To compare the results of 2 $rows
+   *
+   * @param  array $newRow
+   *     a row from a recent view result
+   * @param  array $oldRow
+   *     a row from a previous view result
+   *
+   * @return array
+   *     the difference between both rows
+   */
+  private function diffViewsRow($newRow, $oldRow) {
+    // Normalising new row to get better diffs
+    foreach ($oldRow as $key => $value) {
+      if (!isset($newRow[$key])) {
+        $newRow[$key] = '';
+      }
+    }
+    return array_diff_assoc($newRow, $oldRow);
+  }
+
+  /**
+   * To compare results of 2 views
+   *
+   * @param  array $newData
+   *     a recent result from view
+   * @param  array $oldData
+   *     previous result from view
+   *
+   * @return array
+   *     the difference
+   */
+  private function diffViewsData($newData, $oldData) {
+    $dataDiff = [];
+
+    // Normalising new data to get better diffs
+    foreach ($oldData as $index => $value) {
+      if (!isset($newData[$index])) {
+        $newData[$index] = [];
+      }
+    }
+    // Normalising old data to get better diffs
+    foreach ($newData as $index => $value) {
+      if (!isset($oldData[$index])) {
+        $oldData[$index] = [];
+      }
+    }
+
+    foreach ($oldData as $index => $value) {
+      $diff = $this->diffViewsRow($newData[$index], $oldData[$index]);
+      if ($diff) {
+        $dataDiff[$index] = $diff;
+      }
+    }
+    return $dataDiff;
+  }
+
+  /**
+   * Obtain $this->changes comparing $this->oldData with $this->newData
+   * @param  array $oldData
+   *     previous result from view
+   * @param  array $newData
+   *     a recent result from view
+   * @param  array $changes
+   *     the difference between $oldData and $newData
+   *
+   * @return stdClass $this
+   */
+  private function compareDataRecursive(&$oldData, &$newData, &$changes) {
```

Comment:
Why are $oldData and $newData passed by reference if their values are just checked in this function?